D morozov

DmitryMorozov62 opened the pull request.

DmitryMorozov62 committed:
добавил javadoc

DmitryMorozov62 committed:
Merge branch 'Dev' into DMorozov

DmitryMorozov62 committed:
Merge branch 'Dev' into DMorozov

DmitryMorozov62 committed:
Merge branch 'Dev' into DMorozov

DmitryMorozov62 committed:
работа с изображениями

DmitryMorozov62 committed:
Merge branch 'Dev' into DMorozov

# Conflicts:
#	src/main/java/ru/skypro/homework/service/impl/UserServiceImpl.java

DmitryMorozov62 committed:
exception для user (устронил дубляж)

BizinMitya commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -32,7 +33,11 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+                .antMatchers(HttpMethod.GET,"/login"
+                                                     , "/register"
+                                                     , "/ads"
+                                                     , "/ads/get/*"
+                                                     , "/users/get/*").permitAll()
```
/ads/get - таких путей нет в спецификации

dmitry-bizin commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -32,7 +33,11 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+                                                     , "/ads"
+                                                     , "/ads/get/*"
+                                                     , "/users/get/*").permitAll()
                 .antMatchers("/ads/**", "/users/**").access("hasRole('USER') || hasRole('ADMIN')")
```
какой смысл делать проверку ролей, если их и так всего 2 и обоим разрешено?

dmitry-bizin commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -32,7 +33,11 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                 .authorizeRequests()
-                .antMatchers("/login", "/register", "/ads").permitAll()
+                .antMatchers(HttpMethod.GET,"/login"
+                                                     , "/register"
```
почему GET на /login и /register? На эти пути есть только POST, а не GET

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -95,9 +152,80 @@ public void deletedAd(int idPk, String userName) {
             throw new AdNotFoundException(idPk);
         }
         if (user.getRole().equals(Role.ADMIN) || ad.getUser().getId() == user.getId()) {
+            commentService.deleteAllCommentByPk(idPk);
```
картинку тоже надо удалять из ФС

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -95,9 +152,80 @@ public void deletedAd(int idPk, String userName) {
     }
+
+    @Override
+    @PreAuthorize("hasRole('USER')")
```
почему только юзер-то? админ тоже может

AlexCh773 replied:
Привет! Ответь пожалуйста на вопрос из прошлого пула на эту тему.
![image](https://github.com/yanfeofanov/team8_avito_app/assets/120711675/cbb1107d-4834-448e-807b-a03a134eea66)

dmitry-bizin replied:
Проверка аутентификации сконфигурирована статически в WebSecurityConfig, там написано, что запросы на /ads/** и /users/** должны выполняться только для аутентифицированных. Вот и ответ - если ты выполняешь код в этих методах, значит аутентификация прошла успешно

AlexCh773 replied:
Ошибся немного, совет дублировать проверку в методе сервиса был не в задании а в книге по spring - "Spring in  Action"
![image](https://github.com/yanfeofanov/team8_avito_app/assets/120711675/5cf50fa6-6052-4873-a84c-a794ebd51aca)
![image](https://github.com/yanfeofanov/team8_avito_app/assets/120711675/051ae621-ace8-433b-abf6-b7a68d366485)
Тут глупость написана или такая проверка имеет смысл?
Если имеет и при этом, как ты говоришь, нет смысла проверять наличие роли так как любая из них дает право, то снова тот же вопрос - я могу как-то проверить авторизован ли пользователь с помощью аннотации? (сам такой не нашел)

dmitry-bizin replied:
Надо уметь применять инструменты. В книге написано про то, что один и тот же метод может вызываться в разных контекстах и если во всех контекстах нужна проверка, то её лучше делать на уровне метода. В вашем же проекте контекст один - вызов из контроллера, поэтому статической конфигурации в WebSecurityConfig достаточно. Далее, не надо путать аутентификацию и авторизацию. Про аутентификацию я написал выше. По авторизации всё то же самое - если метод начал своё выполнение и над ним стоит аннотация PreAuthorize, то можно быть уверенным, что метод вызвался авторизованным пользователем

AlexCh773 replied:
ок, но я все равно не понимаю почему плохо сразу добавить такую проверку и избежать возможной проблемы при расширении функционала программы, пусть он возможно никогда и не случится. Она "есть не просит", ресурсов больших не требует. 
Разницу между аутентификацией и авторизацией я понимаю, просто не по-русски что-ли как-то получается "аутентифицирован ли пользователь".    

> По авторизации всё то же самое - если метод начал своё выполнение и над ним стоит аннотация PreAuthorize, то можно быть уверенным, что метод вызвался авторизованным пользователем.

Вот эту часть не понял. Ты же хочешь чтобы мы убрали вообще эту аннотацию, так как проверка излишняя. Или как?

dmitry-bizin replied:
>ок, но я все равно не понимаю почему плохо сразу добавить такую проверку и избежать возможной проблемы при расширении функционала программы, пусть он возможно никогда и не случится. Она "есть не просит", ресурсов больших не требует.

Потому что проверка неправильная. Ещё раз. По заданию ОБОИМ ролям разрешено вызывать эндпоинт, а не только юзеру. Во-вторых, ролей и так всего 2, поэтому если их написать в аннотации, то эффект будет точно такой же, как если бы просто не писать эту аннотацию.

>Вот эту часть не понял. Ты же хочешь чтобы мы убрали вообще эту аннотацию, так как проверка излишняя. Или как?

Да, хочу. Но ты спросил, как проверить, что метод выполняется авторизованным. Через эту аннотацию, например, можно проверить. Но это будет излишне, потому что в WebSecurityConfig это уже настроено и дублировать не нужно.

>ок, но я все равно не понимаю почему плохо сразу добавить такую проверку и избежать возможной проблемы при расширении функционала программы

Почитай про KISS и YAGNI принципы.

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -95,9 +152,80 @@ public void deletedAd(int idPk, String userName) {
+     * @return String
+     */
+    private String getExtension(String fileName) {
+        return fileName.substring(fileName.lastIndexOf(".") + 1);
```
лучше использовать спринговый класс StringUtils и его метод getFilenameExtension

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/AdServiceImpl.java:
```diff
@@ -95,9 +152,80 @@ public void deletedAd(int idPk, String userName) {
+     * @param userEmail login пользователя
+     * @return String
+     */
+    private String getUrlImage(int idPk, MultipartFile image, String userEmail) {
```
метод слишком много делает - создаёт директории, записывает в файл, формирует url, называется get, а на самом деле там не только get - спагетти-код

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/impl/UserServiceImpl.java:
```diff
@@ -54,4 +68,46 @@ private AvitoUser getAuthUser() {
     }
+
+    @Override
+    @PreAuthorize("hasRole('USER')")
```
это неверно, админ тоже может

DmitryMorozov62 committed:
1)Убрал из WebSecurityConfig HttpMethod.GET
2)Удаляю картинку из ФС и про обновновлении удаляю старую
3)использую StringUtils для получения расширения
4)закомментил PreAuthorize
5)Разбил метод getUrlImage

DmitryMorozov62 committed:
1)access -> authenticated
